genienlp/train.py

```diff
@@ -493,12 +503,41 @@ def main(args):
     else:
         logger.info(f'Initializing a new {model_name}')
         model = model_class(args=args, vocab_sets=train_sets+val_sets, tasks=tasks)
+        model.set_decoder_start_token_id(args.train_languages.split('+')[0])
 
     params = get_trainable_params(model)
     log_model_size(logger, model, model_name)
-
-    model.to(devices[0])
-    model = NamedTupleCompatibleDataParallel(model, device_ids=devices)
+    
+    if args.model_parallel:
+        n_layers = len(model.model.encoder.block)
+        layers = list(range(n_layers))
+        
+        if args.mp_device_ratio is not None:
+            if len(args.devices) > torch.cuda.device_count() or max(args.devices) >= torch.cuda.device_count():
+                raise ValueError('Provided GPU devices exceeds the number of available GPU')
+            
+            mp_device_ratio = [(device_ratio / sum(args.mp_device_ratio)) for device_ratio in args.mp_device_ratio]
+            layers_list = []
+            beg = 0
+            for i in range(len(args.devices)):
+                end = min(beg + math.ceil(n_layers * mp_device_ratio[i]), n_layers)
+                layers_list.append(layers[beg: end])
+                beg = end
+                
+            if any([len(val_list) == 0 for val_list in layers_list]):
+                raise ValueError('One device got no layers given the mp_device_ratio you provided'
+                                 'Hint: if you do not provide mp_device_ratio, layers will be distributed evenly')
+            
+        else:
+            n_blocks = int(math.ceil(n_layers / len(args.devices)))
+            layers_list = list(layers[i: i + n_blocks] for i in range(0, n_layers, n_blocks))
+            
+        device_map = dict(zip(args.devices, layers_list))
+        model.model.parallelize(device_map)
```

Comment:
This is TransformerSeq2Seq specific, does it work with TransformerLSTM? Or does it crash miserably?

Reply:
No and only specific models within TransformerSeq2Seq support it so far. I will change it to raise error in those cases.